### PR TITLE
Archive after editing doesn't save the note #11

### LINKED ISSUE #11: Archive after editing doesn't save the note

### PR DESCRIPTION
**Pull request by faruktoptas**
- Fix #11 
- Merge master to develop
- Fix author name

### CODE REVIEW

**Comment by vishesh** [comment]
Hello @faruktoptas, could you please rebase it again against the develop branch. I've merged master with develop myself.

**Comment by faruktoptas** [comment]
I rebased @vishesh

**Comment by faruktoptas** [comment]
I rebased @vishesh

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java [hunk 1]
```diff
@@ -490,9 +507,13 @@ public void doUnarchive() {
...
 
     /**
-     * Archive current note
+     * Save not if edited then Archive current note
      */
     public void doArchive() {
+        if (mNoteChanged){
+            saveNote();
+        }
+        saveNote();
```
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: This second save seems unnecessary.

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java [hunk 2]
```diff
@@ -551,6 +572,7 @@ public void onColorChanged(int color) {
             colorStrip1.setBackgroundColor(Misc.getColorForCode(this, color));
             colorStrip1.setVisibility(View.VISIBLE);
         }
+        mNoteChanged = true;
```
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: Archiving ends current note activity, so this is unnecessary.

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java [hunk 3]
```diff
@@ -237,6 +255,7 @@ private void loadNote(Note note) {
             mTitleView.setEnabled(false);
             ((View) mNoteView).setEnabled(false);
         }
+        mNoteChanged = false;
```
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: [CR0] Initialize mNoteChange at declaration.

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java [hunk 4]
```diff
@@ -53,6 +53,7 @@
     private TagEditText mTagEditText;    /* Chips view editor for tags */
 
     private int         mBackgroundColor;
+    private boolean     mNoteChanged;
```
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: See [CR0]. Initialize to false.

On other hand I think, we don't need `mNoteChanged` at first place. We already know the original state of note in object `mNote` which is either `null` for new notes and loaded via `loadNote` for saved notes. You can add a predicate function `isNoteChanged` which can compare current state with original state. This way, you get less state to manage which is a good thing.

**Review by vishesh — COMMENTED** [review]
Thanks @faruktoptas for fixing it. I've reviewed your code, as well as suggested another way to just compute things out of existing state, rather than managing more.

Also, try to split unrelated changes (config files + feature/bug) in different commit unless build is failing without other. In that case, just add couple lines in commit message regarding that.

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java (hunk 4 above)
Comment (faruktoptas) [reply Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: I removed `mNoteChanged`, it is unnecessary. In `saveNote()` method it is checked that if `mNote` is changed ( `final boolean anythingChanged` ).  But it saves the note if `mAutoSaveEnabled` is `true`. I will add a `forceSave` parameter to the `save()` method it will save the note without enabling auto-save. Is this ok?

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java (hunk 3 above)
Comment (faruktoptas) [reply Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: It will be removed.

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java (hunk 1 above)
Comment (faruktoptas) [reply Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: Sorry, I forgot it there :)

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java (hunk 2 above)
Comment (faruktoptas) [reply Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: It will be removed.

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java (hunk 4 above)
Comment (vishesh) [reply Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: Yes. Sounds good.

Comment (akshatgoel) [reply Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: @faruktoptas @vishesh Sorry for chiming in late but I guess no damage is done :)

I will play devil's advocate here and will propose that we don't take decision on user behalf. Why not ask user if he wants to save before archiving? This will simplify the flow and also, save accidental "archive" instead of "saves".

A simple "Your note has changed. Do you want to <save and archive> or <don't archive>" can warn user that saving != archiving.

Comment (vishesh) [reply Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: ^+1

Comment (faruktoptas) [reply Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: I agree with you. Maybe we can ask "You have unsaved changes. Do you want to save your note  before archiving? "

Comment (faruktoptas) [reply Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: So we need to separate `isContentChanged()` and `isNoteChanged()` methods out of `doSave()` method. Then the alert dialog will be shown if note is changed. By separating this methods we won't repeat the code to check note changes.

Comment (vishesh) [reply Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: I often mistap on archive when I originally mean save. In that case use would like to be asked if I would like to cancel the op as well. So something like "You have unsaved changes? Do you want to save note before archivnig? Yes, No, Cancel"

Why do we need `isNoteChanged()` and `isContentChanged()`. Shouldn't these be same?  What behavior is different content change as compared to metadata change?

Comment (faruktoptas) [reply Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: I will change the alert dialog like that.

The need for `isContentChanged()` is [here](https://github.com/vishesh/sealnote/blob/master/Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java#L462)

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java [hunk 5]
```diff
@@ -28,39 +31,39 @@
...
-    private Intent      mShareIntent;
-    private boolean     mSaveButtonClicked = false; /* To avoid action on reclicking  */
-    private boolean     mAutoSaveEnabled;
-    private boolean     mLoadingNote = true;    /* Is note loaded from database? */
+    private Note mNote;
+    private Note.Type mNoteType;
+    private Intent mShareIntent;
+    private boolean mSaveButtonClicked = false; /* To avoid action on reclicking  */
+    private boolean mAutoSaveEnabled;
+    private boolean mLoadingNote = true;    /* Is note loaded from database? */
```
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: [CR0] Our style guide likes horizontal alignment at class level at-least. Regardless, style changes to unrelated code should be different commit for cleaner diffs.

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java [hunk 6]
```diff
@@ -28,39 +31,39 @@
...
-    private TextView    mEditedView;    /* Show last edited date/time */
-    private NoteView    mNoteView;
+    private EditText mTitleView;
+    private ViewStub mContentStub;   /* Inflated by appropriate view to edit given note type */
+    private TextView mEditedView;    /* Show last edited date/time */
+    private NoteView mNoteView;
     private TagEditText mTagEditText;    /* Chips view editor for tags */
 
-    private int         mBackgroundColor;
+    private int mBackgroundColor;
```
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: Refer [CR0].

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java [hunk 7]
```diff
@@ -497,9 +517,20 @@ public void doUnarchive() {
     }
 
     /**
-     * Archive current note
+     * Ask user if note is changed, then Archive current note
      */
     public void doArchive() {
+        if (isNoteChanged()){
+            showArchiveAlert();
+        }else {
```
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: Space before else.

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java [hunk 8]
```diff
@@ -497,9 +517,20 @@ public void doUnarchive() {
     }
 
     /**
-     * Archive current note
+     * Ask user if note is changed, then Archive current note
      */
     public void doArchive() {
+        if (isNoteChanged()){
```
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: Space before `{`

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java [hunk 9]
```diff
@@ -513,6 +544,24 @@ public void doArchive() {
     }
 
     /**
+     * Shows an alert before archiving if note is changed.
+     */
+    private void showArchiveAlert(){
```
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: Space before `{`

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java [hunk 10]
```diff
@@ -532,8 +581,9 @@ public void doRestore() {
      * Called when save action is invoked by click
      */
     public void doSave() {
-        if (mSaveButtonClicked) return; else mSaveButtonClicked = true; //FIXME: Hack. Avoids double saving
-        final boolean noteSaved = saveNote();
+        if (mSaveButtonClicked) return;
+        else mSaveButtonClicked = true; //FIXME: Hack. Avoids double saving
+        final boolean noteSaved = saveNote(false);
```
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: Since `noteSaved` is used just once you can inline it. I wrote the original code, but since you are changing these lines, lets clean this little part as well.

**Review by vishesh — APPROVED** [review]
LG to me except for some minor style comments.

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java (hunk 5 above)
Comment (faruktoptas) [reply Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: My Android Studio configuration :)

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java (hunk 10 above)
Comment (faruktoptas) [reply Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: I fixed them all, and pushed. Also removed pushed `.iml` files.

**Comment by vishesh** [comment]
Merged. Thanks @faruktoptas !

CLOSES #11
